superdesk/activity.py

```diff
@@ -147,23 +156,38 @@ def on_update(self, updates, original):
         user_id = str(user.get('_id'))
 
         # make sure that the user making the read notification is in the notification list
-        if user_id not in updates.get('read').keys():
+        if not self.is_recipient(updates, user_id):
             raise SuperdeskApiError.forbiddenError('User is not in the notification list')
 
         # make sure the transition is from not read to read
-        if not (updates.get('read')[user_id] == 1 and original.get('read')[user_id] == 0):
+        if not self.is_read(updates, user_id) and self.is_read(original, user_id):
             raise SuperdeskApiError.forbiddenError('Can not set notification as read')
 
         # make sure that no other users are being marked as read
-        for read_entry in updates.get('read'):
-            if read_entry != user_id:
-                if updates.get('read')[read_entry] != original.get('read')[read_entry]:
+        for recipient in updates.get('recipients'):
+            if recipient['user_id'] != user_id:
+                if self.is_read(updates, user_id) != self.is_read(original, user_id):
                     raise SuperdeskApiError.forbiddenError('Can not set other users notification as read')
 
         # make sure that no other fields are being up dated just read and _updated
         if len(updates) != 2:
             raise SuperdeskApiError.forbiddenError('Can not update')
 
+    def is_recipient(self, activity, user_id):
+        """
+        Checks if the given user is in the list of recipients
+        """
+        return any(r for r in activity.get('recipients', []) if r['user'] == user_id)
+
+    def is_read(self, activity, user_id):
+        """
+        Returns the read value for the given user
+        """
+        try:
+            return next(r['read'] for r in activity.get('recipients', []) if r['user'] == user_id)
+        except StopIteration:
```

Comment:
Let's just pass `False` as the second argument to `next()` avoiding the need for catching  the `StopIteration` error.
Returning `False` is IMO better than raising an error, at least that's what I would expect based on the method name. On the other hand, if we do want to raise an error, it should be documented in the docstring, and maybe the method renamed to `check_is_read()` or something.